# JCalc quickstart

## This notebook explains possible ways of running JCalc and shows examples of input files so you can run on your own.

### Index file (flag '-n')

The index file is needed so jcalc can understand which are the atoms that represents the dihedrals where the vicinal coupling constant will be calculated.
Below we have an example of what is expected inside the index file, presented as a table. 
The items are separated by tab (\t), where each row is a J value and the columns the expected atoms in the dihedral.

<style>
td {
  font-size: 50px
}
</style>
| HX | CX | CY | HY |
| :----: | :----: | :----: | :----: |
| 255 | 254 | 252 | 253 |
| 255 | 254 | 252 | 253 |


Below we have an example of the contents on the index file, where each line is a J value, with 4 numbers delimited by tab (\t)

## Function used to calculate vicinal coupling constant (J):
$j$$_o$$^-$$^1$ $.$ $^3$J$_H$$_,$$_H$  $=$ $13.86$ $cos$$^2$ $\phi$ $-$ $0.81$ $cos$ $\phi$ + 
$\sum_{i=1}^{I}$ $\Delta$$\chi$$_i$[$0.56$ $-$ $2.32$ $cos$ $^2$ ($\xi$$_i$$\phi$$_o$$^-$$^1$$\phi$ $+$ $17.9$ |$\Delta$$\chi$$_i$|)]

In [1]:
def print_file(file):
    with open(file, "r") as f:
        [print(line) for line in f.readlines()]

## Running JCalc locally
Installing instructions on the README

In [2]:
!jcalc -h

usage: jcalc [-h] [-x XTC] [-t TPR] [-suf SUFFIX] [-skip SKIP] [-n J_INPUT]
             [-hydro] [-p PDB] [-i INPUT_DIR]

optional arguments:
  -h, --help            show this help message and exit
  -x XTC, -xtc XTC, -xtc_filename XTC
                        .xtc filename from molecular dynamics you want to
                        analyze (Needs to be in current directory)
  -t TPR, -tpr TPR, -tpr_filename TPR
                        .tpr filename from molecular dynamics you want to
                        analyze (Needs to be in current directory)
  -suf SUFFIX, -suffix SUFFIX
                        Suffix of analysis, used when running multiple 3JH,H
                        calculations in same directory
  -skip SKIP            Number of frames you want to skip when calcu lating
                        3JH,H values thorugh molecular dynamics
  -n J_INPUT, -ndx J_INPUT
                        J input filename to know which coupling constants will
                        be calculat

## Running on single PDB structure

In this results, we want to calculate the J values from a monosaccharide, choosing two dihedrals:

JH1,H2, being the dihedral: H1 - C1 - C2 - H2;

JH2,H3, being the dihedral: H2 - C2 - C3 - H3.

Below, the index file used (remember, delimited by tab):


In [3]:
!jcalc -p ../test/monosac.pdb -n ../test/j_input.tsv

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'COMPND' at line 5
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'AUTHOR' at line 6
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'CRYST1' at line 7
  PDBConstructionWarning,


Now, we can check the output file, called "monosac_J_values.tsv"

In [4]:
print_file("../test/monosac_J_values.tsv")

24,20	8.66

20,19	9.58



We see that, in this static structure, the J values found are:

**JH1,H2: 8.66**

**JH2,H3: 9.58**

It also returns a log file, where you can monitore the execution of JCalc

In [5]:
print_file("jcalc.log")

07/17/2021 14:50:52 - Starting JCalc for PDB file ../test/monosac.pdb

07/17/2021 14:50:52 - J input file: ../test/j_input.tsv

07/17/2021 14:50:52 - Output file path: /home/joao/laboratorio/jcalc/test/monosac_J_values.tsv

07/17/2021 14:50:52 - Ended JCalc analysis for PDB file



## Running on multiple PDB files

Now, we are going to calculate the J value from a protein dihedral through multiple PDB files created by a molecular dynamics.  

In [6]:
!jcalc -i ../test/frames_pdb -n ../test/j_frames_pdb.tsv

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 279
  PDBConstructionWarning,


This will create a statistical file with the mean J value through the dynamics and its standard deviation.

It will also return the brute values for each frame, separing it for all J values calculated (in this case, only one).

In [7]:
print_file("jcalc.log")

07/17/2021 14:50:53 - Starting JCalc

07/17/2021 14:50:53 - Input directory file: /home/joao/laboratorio/jcalc/test/frames_pdb

07/17/2021 14:50:53 - J input file: ../test/j_frames_pdb.tsv

07/17/2021 14:50:53 - Calculating J values from MD

07/17/2021 14:50:54 - Calculating Statistics from J values

07/17/2021 14:50:54 - Writing J statistics resuls, path: /home/joao/laboratorio/jcalc/examples/statistical_results.txt

07/17/2021 14:50:54 - Writing J 255,253 values through MD, path: /home/joao/laboratorio/jcalc/examples/255,253_values.tsv



In [8]:
print_file("statistical_results.txt")

255,253_mean:	2.88

255,253_stdev:	1.67



In [9]:
print_file("255,253_values.tsv")

frame_0.pdb	1.55

frame_1.pdb	2.27

frame_2.pdb	2.04

frame_3.pdb	2.43

frame_4.pdb	4.89

frame_5.pdb	1.47

frame_6.pdb	1.06

frame_7.pdb	2.35

frame_8.pdb	2.71

frame_9.pdb	1.0

frame_10.pdb	2.0

frame_11.pdb	4.92

frame_12.pdb	4.08

frame_13.pdb	2.64

frame_14.pdb	2.72

frame_15.pdb	2.57

frame_16.pdb	1.22

frame_17.pdb	2.82

frame_18.pdb	2.22

frame_19.pdb	3.91

frame_20.pdb	2.43

frame_21.pdb	4.67

frame_22.pdb	2.2

frame_23.pdb	1.69

frame_24.pdb	3.44

frame_25.pdb	1.58

frame_26.pdb	2.99

frame_27.pdb	1.62

frame_28.pdb	1.0

frame_29.pdb	2.43

frame_30.pdb	1.79

frame_31.pdb	3.66

frame_32.pdb	2.97

frame_33.pdb	2.05

frame_34.pdb	1.85

frame_35.pdb	3.3

frame_36.pdb	2.46

frame_37.pdb	1.9

frame_38.pdb	1.45

frame_39.pdb	2.45

frame_40.pdb	1.6

frame_41.pdb	3.43

frame_42.pdb	10.55

frame_43.pdb	2.65

frame_44.pdb	1.82

frame_45.pdb	3.48

frame_46.pdb	2.29

frame_47.pdb	2.6

frame_48.pdb	3.34

frame_49.pdb	1.82

frame_50.pdb	2.21

frame_51.pdb	3.96

frame_52.pdb	2.53

frame_53.p

## Running on XTC file from GROMACS 

Now, we are going to calculate the J value from a protein dihedral through multiple PDB files created by a molecular dynamics.

In [10]:
!jcalc -x ../../peptide_j_values/2ovn_dm.xtc -t ../../peptide_j_values/2ovn_dm.tpr -suf 1 -skip 1000 -n ../../peptide_j_values/j_input.tsv

                     :-) GROMACS - gmx trjconv, 2018.1 (-:

                            GROMACS is written by:
     Emile Apol      Rossen Apostolov      Paul Bauer     Herman J.C. Berendsen
    Par Bjelkmar    Aldert van Buuren   Rudi van Drunen     Anton Feenstra  
  Gerrit Groenhof    Aleksei Iupinov   Christoph Junghans   Anca Hamuraru   
 Vincent Hindriksen Dimitrios Karkoulis    Peter Kasson        Jiri Kraus    
  Carsten Kutzner      Per Larsson      Justin A. Lemkul    Viveca Lindahl  
  Magnus Lundborg   Pieter Meulenhoff    Erik Marklund      Teemu Murtola   
    Szilard Pall       Sander Pronk      Roland Schulz     Alexey Shvetsov  
   Michael Shirts     Alfons Sijbers     Peter Tieleman    Teemu Virolainen 
 Christian Wennberg    Maarten Wolf   
                           and the project leaders:
        Mark Abraham, Berk Hess, Erik Lindahl, and David van der Spoel

Copyright (c) 1991-2000, University of Groningen, The Netherlands.
Copyright (c) 2001-2017, The GROMACS de

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 

/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 290
  PDBConstructionWarning,
/home/joao/.venv/jcalc/lib/python3.7/site-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 

These are the same frames we analyzed in the "Runing on multiple PDB files" session, however, we extract it automatically using jcalc.

In [11]:
print_file("jcalc.log")

07/17/2021 14:50:54 - Starting JCalc

07/17/2021 14:50:54 - XTC chosen file: ../../peptide_j_values/2ovn_dm.xtc

07/17/2021 14:50:54 - TPR chosen file: ../../peptide_j_values/2ovn_dm.tpr

07/17/2021 14:50:54 - Skip chosen: 1000

07/17/2021 14:50:54 - J input file: ../../peptide_j_values/j_input.tsv

07/17/2021 14:50:54 - Creating frames, path: /home/joao/laboratorio/jcalc/examples/frames1

07/17/2021 14:51:41 - Calculating J values from MD

07/17/2021 14:51:42 - Calculating Statistics from J values

07/17/2021 14:51:42 - Writing J statistics resuls, path: /home/joao/laboratorio/jcalc/examples/statistical_results.txt

07/17/2021 14:51:42 - Writing J 255,253 values through MD, path: /home/joao/laboratorio/jcalc/examples/255,253_values.tsv



In [12]:
print_file("statistical_results.txt")

255,253_mean:	2.88

255,253_stdev:	1.67



In [13]:
print_file("255,253_values.tsv")

frame_0.pdb	1.55

frame_1.pdb	2.27

frame_2.pdb	2.04

frame_3.pdb	2.43

frame_4.pdb	4.89

frame_5.pdb	1.47

frame_6.pdb	1.06

frame_7.pdb	2.35

frame_8.pdb	2.71

frame_9.pdb	1.0

frame_10.pdb	2.0

frame_11.pdb	4.92

frame_12.pdb	4.08

frame_13.pdb	2.64

frame_14.pdb	2.72

frame_15.pdb	2.57

frame_16.pdb	1.22

frame_17.pdb	2.82

frame_18.pdb	2.22

frame_19.pdb	3.91

frame_20.pdb	2.43

frame_21.pdb	4.67

frame_22.pdb	2.2

frame_23.pdb	1.69

frame_24.pdb	3.44

frame_25.pdb	1.58

frame_26.pdb	2.99

frame_27.pdb	1.62

frame_28.pdb	1.0

frame_29.pdb	2.43

frame_30.pdb	1.79

frame_31.pdb	3.66

frame_32.pdb	2.97

frame_33.pdb	2.05

frame_34.pdb	1.85

frame_35.pdb	3.3

frame_36.pdb	2.46

frame_37.pdb	1.9

frame_38.pdb	1.45

frame_39.pdb	2.45

frame_40.pdb	1.6

frame_41.pdb	3.43

frame_42.pdb	10.55

frame_43.pdb	2.65

frame_44.pdb	1.82

frame_45.pdb	3.48

frame_46.pdb	2.29

frame_47.pdb	2.6

frame_48.pdb	3.34

frame_49.pdb	1.82

frame_50.pdb	2.21

frame_51.pdb	3.96

frame_52.pdb	2.53

frame_53.p

## Running on docker

You can run every session as you did before on Docker, just changing some of the sintax. 

**The files that are going to be used in the calculation need to be on your current directory (pwd).**

Here is an example of the single PDB session on docker:

In [14]:
!docker run -v $(pwd):/home/data jlmeirelles/jcalc -p monosac.pdb -n j_input.tsv

/usr/local/lib/python3.6/dist-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'COMPND' at line 5
  PDBConstructionWarning,
/usr/local/lib/python3.6/dist-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'AUTHOR' at line 6
  PDBConstructionWarning,
/usr/local/lib/python3.6/dist-packages/Bio/PDB/PDBParser.py:399: PDBConstructionWarning: Ignoring unrecognized record 'CRYST1' at line 7
  PDBConstructionWarning,


In [15]:
print_file("jcalc.log")

07/17/2021 17:51:45 - Starting JCalc for PDB file monosac.pdb

07/17/2021 17:51:45 - J input file: j_input.tsv

07/17/2021 17:51:45 - Output file path: /home/data/monosac_J_values.tsv

07/17/2021 17:51:45 - Ended JCalc analysis for PDB file



In [16]:
print_file("monosac_J_values.tsv")

24,20	8.66

20,19	9.58

